Invalid usage of cv2.copyMakeBorder in post_dark_udp

**Describe the bug**

Evaluation of models with udp is broke

**Reproduction**

- What command or script did you run?

```
python3 tools/test.py configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/hrnet_w32_coco_256x192_udp.py https://download.openmmlab.com/mmpose/top_down/udp/hrnet_w32_coco_256x192_udp-aba0be42_20210220.pth
```

**Environment**

sys.platform: linux
Python: 3.8.5 (default, Jun 24 2021, 12:45:07) [GCC 9.3.0]
CUDA available: True
GPU 0: GeForce RTX 3060
CUDA_HOME: None
GCC: gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
PyTorch: 1.10.1+cu113
PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191122 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.2.3 (Git Hash 7336ca9f055cf1bfa13efb658fe15dc9b41f0740)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.3
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86
  - CuDNN 8.2
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.3, CUDNN_VERSION=8.2.0, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.10.1, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, 

TorchVision: 0.11.2+cu113
OpenCV: 4.5.4
MMCV: 1.3.18
MMCV Compiler: GCC 7.3
MMCV CUDA Compiler: 11.3
MMPose: 0.20.0+59c5cc1

**Error traceback**

If applicable, paste the error traceback here.

```
loading annotations into memory...
Done (t=0.26s)
creating index...
index created!
=> Total boxes: 104125
=> Total boxes after filter low score@0.0: 104125
=> num_images: 5000
=> load 104125 samples
load checkpoint from http path: https://download.openmmlab.com/mmpose/top_down/udp/hrnet_w32_coco_256x192_udp-aba0be42_20210220.pth
Downloading: "https://download.openmmlab.com/mmpose/top_down/udp/hrnet_w32_coco_256x192_udp-aba0be42_20210220.pth" to /home/qt/.cache/torch/hub/checkpoints/hrnet_w32_coco_256x192_udp-aba0be42_20210220.pth
100%|█████████████████████████████████████████████████████████████████████████████████| 109M/109M [00:13<00:00, 8.61MB/s]
[                                                  ] 0/104125, elapsed: 0s, ETA:Traceback (most recent call last):
  File "tools/test.py", line 174, in <module>
    main()
  File "tools/test.py", line 150, in main
    outputs = single_gpu_test(model, data_loader)
  File "/home/qt/.pyenv/versions/mmpose/lib/python3.8/site-packages/mmpose/apis/test.py", line 33, in single_gpu_test
    result = model(return_loss=False, **data)
  File "/home/qt/.pyenv/versions/mmpose/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/qt/.pyenv/versions/mmpose/lib/python3.8/site-packages/mmcv/parallel/data_parallel.py", line 50, in forward
    return super().forward(*inputs, **kwargs)
  File "/home/qt/.pyenv/versions/mmpose/lib/python3.8/site-packages/torch/nn/parallel/data_parallel.py", line 166, in forward
    return self.module(*inputs[0], **kwargs[0])
  File "/home/qt/.pyenv/versions/mmpose/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/qt/.pyenv/versions/mmpose/lib/python3.8/site-packages/mmcv/runner/fp16_utils.py", line 98, in new_func
    return old_func(*args, **kwargs)
  File "/home/qt/.pyenv/versions/mmpose/lib/python3.8/site-packages/mmpose/models/detectors/top_down.py", line 139, in forward
    return self.forward_test(
  File "/home/qt/.pyenv/versions/mmpose/lib/python3.8/site-packages/mmpose/models/detectors/top_down.py", line 190, in forward_test
    keypoint_result = self.keypoint_head.decode(
  File "/home/qt/.pyenv/versions/mmpose/lib/python3.8/site-packages/mmpose/models/heads/topdown_heatmap_base_head.py", line 74, in decode
    preds, maxvals = keypoints_from_heatmaps(
  File "/home/qt/.pyenv/versions/mmpose/lib/python3.8/site-packages/mmpose/core/evaluation/top_down_eval.py", line 570, in keypoints_from_heatmaps
    preds = post_dark_udp(preds, heatmaps, kernel=kernel)
  File "/home/qt/.pyenv/versions/mmpose/lib/python3.8/site-packages/mmpose/core/evaluation/top_down_eval.py", line 369, in post_dark_udp
    batch_heatmaps_pad = cv2.copyMakeBorder(
cv2.error: OpenCV(4.5.4) /tmp/pip-req-build-24v3eisa/opencv/modules/core/src/copy.cpp:1026: error: (-215:Assertion failed) top >= 0 && bottom >= 0 && left >= 0 && right >= 0 && _src.dims() <= 2 in function 'copyMakeBorder'
```

**Bug fix**

Evaluating model with udp is invalid and threw exception:
https://github.com/open-mmlab/mmpose/blob/master/mmpose/core/evaluation/top_down_eval.py#L369

The function `cv2.copyMakeBorder` [doesn't work with 3D tensors](https://github.com/opencv/opencv/issues/21274), which `batch_heatmaps` is

--- a/mmpose/core/evaluation/top_down_eval.py
+++ b/mmpose/core/evaluation/top_down_eval.py
@@ -366,34 +366,10 @@ def post_dark_udp(coords, batch_heatmaps, kernel=3):
             cv2.GaussianBlur(heatmap, (kernel, kernel), 0, heatmap)
     np.clip(batch_heatmaps, 0.001, 50, batch_heatmaps)
     np.log(batch_heatmaps, batch_heatmaps)
-    batch_heatmaps = np.transpose(batch_heatmaps,
-                                  (2, 3, 0, 1)).reshape(H, W, -1)
 
-    # cv2.copyMakeBorder will report an error when input dimension exceeds 512
-    batch_heatmaps_channel = batch_heatmaps.shape[2]
-    if batch_heatmaps_channel > 512:
-        total_group_number = int(np.ceil(batch_heatmaps_channel / 512))
-        splited_batch_heatmaps = []
-        for group_idx in range(total_group_number):
-            splited_batch_heatmap = batch_heatmaps[
-                ..., group_idx *
-                512:min(batch_heatmaps_channel, (group_idx + 1) * 512)]
-            batch_heatmap_pad = cv2.copyMakeBorder(
-                splited_batch_heatmap,
-                1,
-                1,
-                1,
-                1,
-                borderType=cv2.BORDER_REFLECT)
-            splited_batch_heatmaps.append(batch_heatmap_pad)
-        batch_heatmaps_pad = np.concatenate(splited_batch_heatmaps, axis=2)
-    else:
-        batch_heatmaps_pad = cv2.copyMakeBorder(
-            batch_heatmaps, 1, 1, 1, 1, borderType=cv2.BORDER_REFLECT)
-
-    batch_heatmaps_pad = np.transpose(
-        batch_heatmaps_pad.reshape(H + 2, W + 2, B, K),
-        (2, 3, 0, 1)).flatten()
+    batch_heatmaps_pad = np.pad(
+        batch_heatmaps, ((0, 0), (0, 0), (1, 1), (1, 1)),
+        mode='edge').flatten()
 
     index = coords[..., 0] + 1 + (coords[..., 1] + 1) * (W + 2)
     index += (W + 2) * (H + 2) * np.arange(0, B * K).reshape(-1, K)

